lib/_007/Builtins.pm

```diff
@@ -1,410 +1,416 @@
-    multi equal-value(Val::Int $l, Val::Int $r) { $l.value == $r.value }
-    multi equal-value(Val::Str $l, Val::Str $r) { $l.value eq $r.value }
-    multi equal-value(Val::Array $l, Val::Array $r) {
-        if %*equality-seen{$l.WHICH} && %*equality-seen{$r.WHICH} {
-            return $l === $r;
-        }
-        %*equality-seen{$l.WHICH}++;
-        %*equality-seen{$r.WHICH}++;
-
-        sub equal-at-index($i) {
-            equal-value($l.elements[$i], $r.elements[$i]);
-        }
+sub wrap($_) {
+    when Val | Q { $_ }
+    when Nil  { NONE }
+    when Bool { Val::Bool.new(:value($_)) }
+    when Int  { Val::Int.new(:value($_)) }
+    when Str  { Val::Str.new(:value($_)) }
+    when Array | Seq | List { Val::Array.new(:elements(.map(&wrap))) }
+    default { die "Got some unknown value of type ", .^name }
+}
 
-        [&&] $l.elements == $r.elements,
-            |(^$l.elements).map(&equal-at-index);
+# These multis are used below by infix:<==> and infix:<!=>
+multi equal-value($, $) { False }
+multi equal-value(Val::NoneType, Val::NoneType) { True }
+multi equal-value(Val::Bool $l, Val::Bool $r) { $l.value == $r.value }
+multi equal-value(Val::Int $l, Val::Int $r) { $l.value == $r.value }
+multi equal-value(Val::Str $l, Val::Str $r) { $l.value eq $r.value }
+multi equal-value(Val::Array $l, Val::Array $r) {
+    if %*equality-seen{$l.WHICH} && %*equality-seen{$r.WHICH} {
+        return $l === $r;
     }
-    multi equal-value(Val::Object $l, Val::Object $r) {
-        if %*equality-seen{$l.WHICH} && %*equality-seen{$r.WHICH} {
-            return $l === $r;
-        }
-        %*equality-seen{$l.WHICH}++;
-        %*equality-seen{$r.WHICH}++;
+    %*equality-seen{$l.WHICH}++;
+    %*equality-seen{$r.WHICH}++;
 
-        sub equal-at-key(Str $key) {
-            equal-value($l.properties{$key}, $r.properties{$key});
-        }
-
-        [&&] $l.properties.keys.sort.perl eq $r.properties.keys.sort.perl,
-            |($l.properties.keys).map(&equal-at-key);
-    }
-    multi equal-value(Val::Type $l, Val::Type $r) {
-        $l.type === $r.type
+    sub equal-at-index($i) {
+        equal-value($l.elements[$i], $r.elements[$i]);
     }
-    multi equal-value(Val::Func $l, Val::Func $r) {
-        $l.name eq $r.name
-            && equal-value($l.parameterlist, $r.parameterlist)
-            && equal-value($l.statementlist, $r.statementlist)
-    }
-    multi equal-value(Q $l, Q $r) {
-        sub same-avalue($attr) {
-            equal-value($attr.get_value($l), $attr.get_value($r));
-        }
 
-        [&&] $l.WHAT === $r.WHAT,
-            |$l.attributes.map(&same-avalue);
+    [&&] $l.elements == $r.elements,
+        |(^$l.elements).map(&equal-at-index);
+}
+multi equal-value(Val::Object $l, Val::Object $r) {
+    if %*equality-seen{$l.WHICH} && %*equality-seen{$r.WHICH} {
+        return $l === $r;
     }
+    %*equality-seen{$l.WHICH}++;
+    %*equality-seen{$r.WHICH}++;
```

Comment:
no `LEAVE %*equality-seen{$l.WHICH}--;`?

Reply:
The values in `%*equality-seen` are essentially booleans (if not in actual fact). They start out false and then they flip to true once something has been seen. They don't decrement because they don't flip back; once you have seen something, it can't be unseen. 😄 

By the way, the line you are commenting on is unchanged in this PR save for indentation. If you would like to review code without indentation changes, [the ?w=1 option](https://github.com/masak/007/pull/310/files?w=1) allows you to do that. I find it terribly handy.